types/react/index.d.ts

```diff
@@ -560,6 +560,159 @@ declare namespace React {
+    // specially. I don't think it's intended to accept async functions, should use throw instead of
+    // await to properly engage Suspense.
+    type EffectCallback = () => (void | (() => void));
+
+    interface MutableRefObject<T> {
+        current: T;
+    }
+
+    // This will technically work if you give a Consumer<T> or Provider<T> but it's deprecated and warns
+    /**
+     * Accepts a context object (the value returned from `React.createContext`) and returns the current
+     * context value, as given by the nearest context provider for the given context.
+     *
+     * @version experimental
+     * @see https://reactjs.org/docs/hooks-reference.html#usecontext
+     */
+    function useContext<T>(context: Context<T>/*, (not public API) observedBits?: number|boolean */): T;
+    /**
+     * Returns a stateful value, and a function to update it.
+     *
+     * @version experimental
+     * @see https://reactjs.org/docs/hooks-reference.html#usestate
+     */
+    function useState<S>(initialState: S | (() => S)): [S, Dispatch<SetStateAction<S>>];
+    /**
+     * An alternative to `useState`.
+     *
+     * `useReducer` is usually preferable to `useState` when you have complex state logic that involves
+     * multiple sub-values. It also lets you optimize performance for components that trigger deep
+     * updates because you can pass `dispatch` down instead of callbacks.
+     *
+     * @version experimental
+     * @see https://reactjs.org/docs/hooks-reference.html#usereducer
+     */
+    function useReducer<S, A>(reducer: Reducer<S, A>, initialState: S, initialAction?: A | null): [S, Dispatch<A>];
+    /**
+     * `useRef` returns a mutable ref object whose `.current` property is initialized to the passed argument
+     * (`initialValue`). The returned object will persist for the full lifetime of the component.
+     *
+     * Note that `useRef()` is useful for more than the `ref` attribute. It’s handy for keeping any mutable
+     * value around similar to how you’d use instance fields in classes.
+     *
+     * @version experimental
+     * @see https://reactjs.org/docs/hooks-reference.html#useref
+     */
+    function useRef<T>(initialValue: T): MutableRefObject<T>;
+    // convenience overload for refs given as a ref prop as they typically start with a null value
+    /**
+     * `useRef` returns a mutable ref object whose `.current` property is initialized to the passed argument
+     * (`initialValue`). The returned object will persist for the full lifetime of the component.
+     *
+     * Note that `useRef()` is useful for more than the `ref` attribute. It’s handy for keeping any mutable
+     * value around similar to how you’d use instance fields in classes.
+     *
+     * @version experimental
+     * @see https://reactjs.org/docs/hooks-reference.html#useref
+     */
+    function useRef<T>(initialValue: T|null): MutableRefObject<T|null>;
```

Comment:
This really just saves you from having to write the `|null` in the generic argument yourself.

Reply:
With `createRef` I see it more as "happens to be mutable" instead of something you're supposed to mutate. Only React itself mutates it normally.

(unpopular?) idea: make this convenience overload return a readonly RefObject since it's explicitly intended to be used for things given as `ref`, where only React should reassign its value

Reply:
I know. This idea is for when you _are_ using `useRef` as the `ref` attribute. If you are using it for component-local storage it'll still go to the first overload that is mutable.